### ISSUE TITLE
Documentation out of date WRT App Engine requests support

### ISSUE THREAD

**Issue by kkinder** [stormpath/stormpath-flask#38]
Just reading the [latest docs](http://flask-stormpath.readthedocs.org/en/latest/).

> Note
>
> Unfortunately, this library will NOT work on Google App Engine due to the requests module not working :(

Fortunately [that issue](https://github.com/kennethreitz/requests/issues/498) has long since been closed and requests does generally work now in GAE. I'm not sure if there are any other issues with using stormpath on GAE.

**Comment by rdegges** [comment]
I tried this about 2 months ago, but it still wasn't working. There's another open issue somewhere talking about it... But I can't seem to find it during a quick skim.

Are you using GAE? If so, wanna help test? I'd be happy to remove the warning if it is indeed working now =)

**Comment by kkinder** [comment]
So, with much wrangling, I was able to get it to work. Here's what I had to do.

First, your call to platform() in http.py doesn't inside the GAE sandbox. That was simple enough to fix; I just check if it's a GAE environment and imagine that up as another OS:

    if 'APPENGINE_RUNTIME' in os.environ.keys():
        USER_AGENT = 'stormpath-sdk-python/%s python/%s Google App Engine' % \
            (STORMPATH_VERSION, '%s.%s.%s' % (vi.major, vi.minor, vi.micro))
    else:
        os_info = platform()
        os_versions = {
            'Linux': "%s (%s)" % (linux_distribution()[0], os_info),
            'Windows': "%s (%s)" % (win32_ver()[0], os_info),
            'Darwin': "%s (%s)" % (mac_ver()[0], os_info),
        }

        USER_AGENT = 'stormpath-sdk-python/%s python/%s %s/%s' % (
            STORMPATH_VERSION,
            '%s.%s.%s' % (vi.major, vi.minor, vi.micro),
            system(),
            os_versions.get(system(), ''),
        )

Then, you need to enable open_ssl in your app.yaml file:

    libraries:
    - name: ssl
      version: latest

Now the ugly part. To bypass an [SDK bug](http://stackoverflow.com/questions/29783483/how-can-i-tell-the-code-is-running-on-google-app-engine/29783627#29783627), you have to whitelist the _ssl and _socket libraries, then monkey-patch the Python socket library over the SDK one.

That's about it.

Having said that, all of that seems a bit ... unnecessary given that there's a very nice http fetch service built into GAE, which would be preferable to use.

**Comment by rdegges** [comment]
Ouch!

So, this is essentially what I bumped into the last time I tried this.

As the library maintainer, it's sorta a difficult call, but going through all this trouble to maintain compatibility for a service which is breaking HTTP requests on purpose is sorta... odd. It feels wrong to include this stuff in the core library here as it involves a lot of 'hackery' which makes things messy.

Thanks so much for going through all the trouble above though. I think that at the very least, any future user who wants to make this work on GAE can reference this and hopefully take it from there.

**Comment by kkinder** [comment]
Yeah, Google App Engine is a weird asterisks in the world of Python. I don't know the details, but originally Google said basically "no sockets." Now they said "yes, sockets" but they replace the default sockets library with their own, which plays nicer with their sandbox.

So yeah, it's a little rough... And weird. But GAE is a very nice little SDK if you can live within its limits, and it would be nice if Stormpath could run inside it. Your actual use of the requests module is limited to only a few lines, and for that matter, you could just as easily use urllib which GAE does a better job of supporting.

I'll send over a patch that makes it work, hacky as it is.

**Comment by kkinder** [comment]
[Relevant pull request](https://github.com/stormpath/stormpath-sdk-python/pull/162).
